syncer/syncer.go

```diff
@@ -718,6 +731,62 @@ func (s *Syncer) addCount(isFinished bool, queueBucket string, tp opType, n int6
 	}
 }
 
+func (s *Syncer) jobsIsEmpty() bool {
+	for idx := range s.jobs {
+		if len(s.jobs[idx]) > 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// updateReplicationLag calculates syncer's replication lag by job, it called after every batch dml job / one skip job / one ddl
+// job is flushed to target db.
+func (s *Syncer) updateReplicationLag(job *job, queueBucketName string) {
+	// when job is nil mean no job in this bucket, need do reset this bucket lag to 0
+	if job == nil {
+		s.workerLagMap[queueBucketName].Store(0)
+		// when job is nill and s.jobs is empty, means all event is consumed, we update lag to 0
+		if s.jobsIsEmpty() {
+			replicationLagGauge.WithLabelValues(s.cfg.Name).Set(float64(0))
+			s.secondsBehindMaster.Store(0)
+		}
+		return
+	}
+
+	failpoint.Inject("BlockSyncerUpdateLag", func(v failpoint.Value) {
+		args := strings.Split(v.(string), ",")
+		jtp := args[0]                // job type
+		t, _ := strconv.Atoi(args[1]) // sleep time
+		if job.tp.String() == jtp {
+			s.tctx.L().Info("BlockSyncerUpdateLag", zap.String("job type", jtp), zap.Int("sleep time", t))
+			time.Sleep(time.Second * time.Duration(t))
+		}
+	})
+
+	var lag int64
+	switch job.tp {
+	case ddl:
+		// NOTE: we handle ddl job separately because ddl job will clean all the dml job before execution
+		lag = time.Now().Unix() - s.tsOffset.Load() - int64(job.eventHeader.Timestamp)
```

Comment:
`time.Now().Unix() - s.tsOffset.Load() - int64(job.eventHeader.Timestamp)` appears three times. Should we use a function to calculate this?

Reply:
added